src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -1249,13 +1253,19 @@ int kds_del_context(struct kds_sched *kds, struct kds_client *client,
 	int i;
 
 	BUG_ON(!mutex_is_locked(&client->lock));
+	if (!cu_ctx->ref_cnt) {
+		kds_err(client, "No opening CU Context");
+		return -EINVAL;
+	}
+
+	/* Decrement the ref count and check whether any existing context
+	 * present. If no active context exists then delete that cu context.
+	 */
+	if (--cu_ctx->ref_cnt)
+		return -EBUSY;
```

Comment:
@saifuddin-xilinx , if ref_cnt is not zero, you can just return 0 instead of -EBUSY.

If you return 0, in xocl_del_context() and xocl_close_cu_context(), you don't need to convert -EBUSY to 0. Please avoid this kind of error code change. It is confusing and not easy to understand. For most of the time, it can be avoided. Usually, if we have to do that, put a comment to explain why it has to be done.

In kds_fini_hw_ctx_client() and _kds_fini_client(), you can use below code.
``` c
                while (cu_ctx->ref_cnt)
			kds_del_context(kds, client, cu_ctx);
```

Usually, kds_del_context will failed in the open context ioctl() code path. User passed in cu index, it might failed and the caller needs to check the return. But when close client and close hwctx, our code is fulling controling. It should not fail. If it failed, that should be a BUG that we need to fix.

Reply:
Thanks Min for the suggestion. Now changes are looking more simpler. Please take a look once again.